vowpalwabbit/core/include/vw/core/array_parameters_dense.h

```diff
@@ -154,10 +153,9 @@ class dense_parameters
 #endif
 
 private:
-  VW::weight* _begin;
+  std::shared_ptr<VW::weight> _begin;
```

Comment:
Why a shared_ptr? Seems more standard to just delete the pointer in the destructor, or at the very least use a unique_ptr (since the copy and assignment operators are deleted)

Reply:
shared_ptr is necessary for shallow_copy() to work, which is used for seed_vw_model

Reply:
ick, i guess seeded models share the weights? We should add a comment about that, because it would cause all kinds of issues if we had multiple models sharing weights in a multi-threaded way

Reply:
Seeded models share both weights and `shared_data`, and both were raw pointers. The base model used to seed other models was responsible for freeing the pointers, and you had to be sure to free that model last. I changed both to use `shared_ptr`, and now they can be freed in any order.

Even worse is daemon mode, which tries to `mmap` the same shared_data across all its child processes, and never `munmap`s the memory. I made that one a `shared_ptr` with custom destructor.

Reply:
Looks like there's already a comment in `vw.cc` about `seed_vw_model` being unsafe

/// Creates a workspace based off of another workspace. What this means is that
/// the model weights and the shared_data object are shared. This function needs
/// to be used with caution. Reduction data is not shared, therefore this
/// function is unsafe to use for situations where reduction state is required
/// for proper operation such as marginal and cb_adf. Learn on a seeded instance
/// is unsafe, and prediction is also potentially unsafe.